crates/bevy_ui/src/geometry.rs

```diff
@@ -0,0 +1,361 @@
+use bevy_math::Vec2;
+use bevy_reflect::Reflect;
+use std::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Sub, SubAssign};
+
+/// A 2-dimensional area defined by a width and height.
+///
+/// It is commonly used to define the size of a text or UI element.
+///
+/// # Example
+///
+/// In this example an UI size gets created. It has a width of 100px and a height of 200px.
+///
+/// ```rust
+/// # use bevy_ui::{Size, Val};
+/// #
+/// let size = Size {
+///     width: Val::Px(100.0),
+///     height: Val::Px(200.0),
+/// };
+///
+/// assert_eq!(size.width, Val::Px(100.0));
+/// assert_eq!(size.height, Val::Px(200.0));
+/// ```
+#[derive(Copy, Clone, PartialEq, Debug, Reflect)]
+#[reflect(PartialEq)]
+pub struct Size<T: Reflect + PartialEq = f32> {
+    /// The width of the 2-dimensional area.
+    pub width: T,
+    /// The height of the 2-dimensional area.
+    pub height: T,
+}
+
+impl<T: Reflect + PartialEq> Size<T> {
+    /// Creates a new [`Size`] from a width and a height.
+    ///
+    /// # Example
+    ///
+    /// In this example an UI size gets created. It has a width of 100px and a height of 200px.
+    ///
+    /// ```rust
+    /// # use bevy_ui::{Size, Val};
+    /// #
+    /// let size = Size::new(Val::Px(100.0), Val::Px(200.0));
+    ///
+    /// assert_eq!(size.width, Val::Px(100.0));
+    /// assert_eq!(size.height, Val::Px(200.0));
+    /// ```
+    pub fn new(width: T, height: T) -> Self {
+        Size { width, height }
+    }
+}
+
+impl<T: Default + Reflect + PartialEq> Default for Size<T> {
+    fn default() -> Self {
+        Self {
+            width: Default::default(),
+            height: Default::default(),
+        }
+    }
+}
+
+/// A type which is commonly used to define positions, margins, paddings and borders.
+///
+/// # Examples
+///
+/// ## Position
+///
+/// A position is used to determine where to place an UI element.
+///
+/// In this example an UI position gets created. It has a left value of 100px and a top value of 50px.
+/// If positioned absolutely this would correspond to an UI element which is positioned 100px to the right
+/// from the left side of the window and 50px down from the top side of the window.
+///
+/// ```rust
+/// # use bevy_ui::{UiRect, Val};
+/// #
+/// let position = UiRect {
+///     left: Val::Px(100.0),
+///     top: Val::Px(50.0),
+///     ..Default::default()
+/// };
+///
+/// assert_eq!(position.left, Val::Px(100.0));
+/// assert_eq!(position.right, Val::Undefined);
+/// assert_eq!(position.top, Val::Px(50.0));
+/// assert_eq!(position.bottom, Val::Undefined);
+/// ```
+///
+/// ## Margin
+///
+/// A margin is used to create space around UI elements, outside of any defined borders.
+///
+/// In this example an UI margin gets created. It has a left value of 10px, a top value of 20px,
```

Comment:
```suggestion
/// In this example we are creating a UI margin. It has a left value of 10px, a top value of 20px,
```

It is hard to read text written in the passive voice in English, so I try to avoid it.